From 0235c89edfcd7da2d31dd06d342ca32730bf82ac Mon Sep 17 00:00:00 2001
From: Mike Frysinger <vapier@chromium.org>
Date: Tue, 29 May 2018 07:00:31 -0400
Subject: [PATCH 20/26] init: use minijail to lock things down a bit

BUG=None
TEST=time is still set, and log output still shows up

Change-Id: Ib66422a2f313afab0e251d54f5c345692d456672
Reviewed-on: https://chromium-review.googlesource.com/1075578
Commit-Ready: Mike Frysinger <vapier@chromium.org>
Tested-by: Mike Frysinger <vapier@chromium.org>
Reviewed-by: Jorge Lucangeli Obes <jorgelo@chromium.org>
---
 init/tlsdated.conf | 34 ++++++++++++++++++++++++++++++++--
 1 file changed, 32 insertions(+), 2 deletions(-)

diff --git a/init/tlsdated.conf b/init/tlsdated.conf
index 90a6ef1..ec45916 100644
--- a/init/tlsdated.conf
+++ b/init/tlsdated.conf
@@ -9,6 +9,8 @@ start on started system-services
 stop on stopping system-services
 respawn
 
+expect fork
+
 # We sync time from Google servers, so use the reduced set of certs that we
 # have specifically for talking to Google systems.
 env GOOGLE_CERTS=/usr/share/chromeos-ca-certificates
@@ -19,7 +21,35 @@ pre-start script
   chmod 0644 /var/cache/tlsdated/timestamp || true
 end script
 
-# When it runs tlsdate, tlsdated stitches together an argument vector for it
+# The various paths we bind mount:
+# - /dev/log: For syslog.
+# - /dev/rtc: To check/sync hwclock.
+# - /run/dbus: To communicate over dbus.
+# - /run/shill: For access to /etc/resolv.conf settings.
+# - /var/cache/tlsdated: Our internal state.
+#
+# We can't enter a new net namespace since this, by design, uses the network.
+#
+# We don't run the main process under seccomp (yet), but tlsdated will run the
+# tlsdate child through a seccomp filter itself before talking to the network.
+#
+# When running tlsdate, tlsdated stitches together an argument vector for it
 # as follows: it begins with everything supplied to it after the --, then
 # appends -H $host -p $port, and maybe -x $proxy if it has a proxy to use.
-exec tlsdated -v -- /usr/bin/tlsdate -v -C "$GOOGLE_CERTS" -l
+script
+  # Not all systems have hardware clocks, so don't fail if they're missing.
+  # tlsdate itself will also automatically skip things if it's missing.
+  RTC_ARG=''
+  if [ -e /dev/rtc ]; then
+    RTC_ARG='-b /dev/rtc'
+  fi
+
+  exec minijail0 -i \
+    -p -v -r --uts -l \
+    --profile minimalistic-mountns -b /dev/log ${RTC_ARG} \
+    -k '/run,/run,tmpfs,MS_NODEV|MS_NOEXEC|MS_NOSUID,mode=755,size=10M' \
+    -b /run/dbus,,1 -b /run/shill \
+    -k '/var,/var,tmpfs,MS_NODEV|MS_NOEXEC|MS_NOSUID,mode=755,size=10M' \
+    -b /var/cache/tlsdated,,1 \
+    /usr/bin/tlsdated -v -- /usr/bin/tlsdate -v -C "${GOOGLE_CERTS}" -l
+end script
-- 
2.20.1

